internal/server/deviceflow.go

```diff
@@ -64,77 +64,112 @@ retry:
+	}
+
+	accessKey := &models.AccessKey{
+		IssuedFor:     user.ID,
+		IssuedForName: user.Name,
+
+		// Share the same provider ID that was used to approve
+		ProviderID:          dfar.ProviderID,
+		ExpiresAt:           time.Now().UTC().Add(a.server.options.SessionDuration),
+		InactivityTimeout:   time.Now().UTC().Add(a.server.options.SessionInactivityTimeout),
+		InactivityExtension: a.server.options.SessionInactivityTimeout,
+		Scopes:              models.CommaSeparatedStrings{models.ScopeAllowCreateAccessKey},
+	}
+
+	bearer, err := data.CreateAccessKey(rctx.DBTxn, accessKey)
+	if err != nil {
+		return nil, fmt.Errorf("%w: creating new access key: %v", internal.ErrUnauthorized, err)
+	}
+
+	user.LastSeenAt = time.Now().UTC()
+	if err := data.UpdateIdentity(rctx.DBTxn, user); err != nil {
+		return nil, fmt.Errorf("%w: update user last seen: %v", internal.ErrUnauthorized, err)
+	}
+
+	a.t.User(accessKey.IssuedFor.String(), user.Name)
+	a.t.OrgMembership(accessKey.OrganizationID.String(), accessKey.IssuedFor.String())
+	a.t.Event("login", accessKey.IssuedFor.String(), accessKey.OrganizationID.String(), Properties{"method": "deviceflow"})
+
+	// Update the request context so that logging middleware can include the userID
+	rctx.Authenticated.User = user
+	c.Set(access.RequestContextKey, rctx)
+
+	org, err := data.GetOrganization(rctx.DBTxn, data.GetOrganizationOptions{ByID: accessKey.OrganizationID})
+	if err != nil {
+		return nil, fmt.Errorf("%w: device flow get organization for user: %v", internal.ErrUnauthorized, err)
+	}
+
+	// Delete the request so it can't be claimed twice
+	err = data.DeleteDeviceFlowAuthRequest(rctx.DBTxn, dfar.ID)
+	if err != nil {
+		return nil, fmt.Errorf("%w: device flow delete auth request: %v", internal.ErrUnauthorized, err)
+	}
+
 	return &api.DeviceFlowStatusResponse{
-		Status:     "pending",
+		Status:     api.DeviceFlowStatusConfirmed,
 		DeviceCode: dfar.DeviceCode,
+		LoginResponse: &api.LoginResponse{
+			UserID:           accessKey.IssuedFor,
+			Name:             accessKey.IssuedForName,
+			AccessKey:        string(bearer),
+			Expires:          api.Time(accessKey.ExpiresAt),
+			OrganizationName: org.Name,
+		},
 	}, nil
 }
 
-const days = 24 * time.Hour
-
-func (a *API) ApproveDeviceAdd(c *gin.Context, req *api.ApproveDeviceFlowRequest) (*api.EmptyResponse, error) {
+func (a *API) ApproveDeviceFlow(c *gin.Context, req *api.ApproveDeviceFlowRequest) (*api.EmptyResponse, error) {
 	rctx := getRequestContext(c)
-	dfar, err := access.FindDeviceFlowAuthRequestForApproval(rctx, strings.Replace(req.UserCode, "-", "", 1))
+
+	if !rctx.Authenticated.AccessKey.Scopes.Includes(models.ScopeAllowCreateAccessKey) {
+		// require the device flow scope to approve access keys provided by login & signup keys, but not
+		// keys resulting from device flow itself
+		return nil, fmt.Errorf("%w: access key missing scope '%s'", internal.ErrUnauthorized, models.ScopeAllowCreateAccessKey)
```

Comment:
Just a note that we end up hiding all of these messages from the user: https://github.com/infrahq/infra/blob/8c6e5f361e561de77c6de330f6c77499d9d54140/internal/server/errors.go#L44-L47

We should find a way to allow us to send messages that we have written explicitly for these errors. I can create an issue for this.

Reply:
Created #3837